pandas/core/indexes/base.py

```diff
@@ -301,11 +301,15 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
               (dtype is not None and is_datetime64_any_dtype(dtype)) or
                 'tz' in kwargs):
             from pandas import DatetimeIndex
-            result = DatetimeIndex(data, copy=copy, name=name,
-                                   dtype=dtype, **kwargs)
+
             if dtype is not None and is_dtype_equal(_o_dtype, dtype):
-                return Index(result.to_pydatetime(), dtype=_o_dtype)
+                # GH#23524 if `data` is already a TZaware DatetimeIndex,
```

Comment:
this seems more like fixing the sympton rather than problem. either that or your comment is not very instructive here.

Reply:
Definitely a problem with the comment.  Passing `dtype=object` to DatetimeIndex _should_ raise.  Anything other than None, datetime64[ns], or DatetimeTZDtype is nonsense in that context.

Reply:
right, so why are you fixing this rather than raising?

Reply:
This is in `Index.__new__`, not `DatetimeIndex.__new__`.  The bug that this is addressing is:

```
dti = pd.date_range('2016-01-01', periods=3, tz='US/Pacific')  # arbitrary tz-aware
pd.Index(dti, dtype=object)   # <-- currently raises incorrectly
```